client/lib/public-key-store.js

```diff
@@ -0,0 +1,101 @@
+var _ = require('lodash');
+var uuid = require('uuid');
+var querystring = require("querystring");
+
+var hashKey = require('./hash-key');
+var getRemotePublicKey = require('./remote-public-key');
+var generateToken = require('./generate-token');
+
+var setCachedPublicKey = function setCachedPublicKey(client, cacheKey, publicKey) {
+    client.log('debug', 'Caching public key for ' + client.config.cert_timeout_sec + 's');
+    client.tokenCache.set(cacheKey, publicKey, client.config.cert_timeout_sec);
+}
+
+var getCachedPublicKey = function getCachedPublicKey(client, cacheKey, refresh, xRequestId, callback) {
+    if (refresh) {
+        return getRemotePublicKey(client, cacheKey, xRequestId, function(err, publicKey) {
+            if (err) {
+                callback(err);
+                return
+            }
+            setCachedPublicKey(client, cacheKey, publicKey);
+            callback(null, publicKey);
+            return;
+        });
+    } else {
+        client.tokenCache.get(cacheKey, function getPublicKeyIfNotInCacheThenVerify(error, publicKey) {
+            if (publicKey && _.isString(publicKey)) {
+                client.log('debug', 'Using public key from cache');
+                callback(null, publicKey);
+                return;
+            }
+
+            return getRemotePublicKey(client, cacheKey, xRequestId, function (err, publicKey) {
+                if (err) {
+                    callback(err);
+                    return
+                }
+                setCachedPublicKey(client, cacheKey, publicKey);
+                callback(null, publicKey);
+                return;
+            });
+        });
+    }
+}
+
+var obtainToken = function obtainToken(client, opts, callback) {
+    var id = opts.id;
+    var secret = opts.secret;
+    var xRequestId = opts.xRequestId || uuid.v4();
+
+    var cacheKey = client._formatCacheKey(
+        "obtain_token:" + hashKey(id)
+    );
+
+    // try cache first
+    client.tokenCache.get(cacheKey, function (err, reply) {
+        if (reply) {
```

Comment:
This is the functional change. Previously we had:

```js
    // try cache first
    this.tokenCache.get(cacheKey, function (err, reply) {
        if (err) {
            callback(err, null);
        } else {
if (reply == null) {```

So now, unless the reply is valid we generate a new token from persona.